src/main/java/org/elasticsearch/action/bulk/TransportShardBulkAction.java

```diff
@@ -122,6 +124,7 @@ protected ShardIterator shards(ClusterState clusterState, InternalRequest reques
 
         long[] preVersions = new long[request.items().length];
         VersionType[] preVersionTypes = new VersionType[request.items().length];
+        Translog.Location location = null;
         for (int requestIndex = 0; requestIndex < request.items().length; requestIndex++) {
```

Comment:
I think this is scary (its this way in another file in this change too). we loop through requests and each one assigns location variable, and only at the end do we processAfter(location). 

So is there an implicit guarantee that location's file will always be the same (some concurrency guarantee) and then from there that location's file pointer is only increasing? If so, it would be great to at least have asserts or hard checks, e.g. assert location == null || (location.file = newLocation.file && location.pointer < newLocatoin.pointer) or whatever this is. Or perhaps there is another way to make it less sneaky.